homeassistant/components/deluge/__init__.py

```diff
@@ -1 +1,356 @@
+    """Set up Deluge from a config entry."""
+    client = DelugeClient(hass, entry)
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
+
+    if not await client.async_setup():
+        return False
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    for platform in PLATFORMS:
+        await hass.config_entries.async_forward_entry_unload(entry, platform)
+
+    client = hass.data[DOMAIN].pop(entry.entry_id)
+    if client.unsub_timer:
+        client.unsub_timer()
+
+    return True
+
+
+async def get_api(hass, entry):
+    """Get Transmission client."""
+    host = entry[CONF_HOST]
+    port = entry[CONF_PORT]
+    username = entry.get(CONF_USERNAME)
+    password = entry.get(CONF_PASSWORD)
+
+    try:
+        api = await hass.async_add_executor_job(
+            DelugeRPCClient, host, port, username, password, True
+        )
+        api.connect()
+        _LOGGER.debug("Successfully connected to %s", host)
+        return api
+
+    except ConnectionRefusedError as error:
+        _LOGGER.error(error)
+        raise CannotConnect
+    except Exception as error:  # pylint: disable=broad-except
+        _LOGGER.error(error)
+        if "Username does not exist" in str(error):
+            raise UserNameError
+        if "Password does not match" in str(error):
+            raise PasswordError
+        raise UnknownError
+
+
+class DelugeClient:
+    """Deluge Client Object."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the Deluge client."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self._deluge_data = None
+        self.unsub_timer = None
+
+    @property
+    def api(self):
+        """Return the _deluge_data object."""
+        return self._deluge_data
+
+    async def async_setup(self):
+        """Set up the Deluge client."""
+
+        try:
+            deluge_api = await get_api(self.hass, self.config_entry.data)
+        except CannotConnect:
+            raise ConfigEntryNotReady
+        except (UserNameError, PasswordError, UnknownError):
+            return False
+
+        self._deluge_data = DelugeData(self.hass, self.config_entry, deluge_api)
+
+        await self.hass.async_add_executor_job(self._deluge_data.init_torrent_list)
+        await self.hass.async_add_executor_job(self._deluge_data.update)
+        self.add_options()
+        self.set_scan_interval(self.config_entry.options[CONF_SCAN_INTERVAL])
+
+        for component in PLATFORMS:
+            self.hass.async_create_task(
+                self.hass.config_entries.async_forward_entry_setup(
+                    self.config_entry, component
+                )
+            )
+
+        def add_torrent(service):
+            """Add new torrent to download."""
+            torrent = service.data[ATTR_TORRENT]
+            deluge_client = None
+            for entry in self.hass.config_entries.async_entries(DOMAIN):
```

Comment:
You're calling an async method here inside a sync context.

Reply:
This is a method called by the `add_torrent` service. This method includes call methods to the deluge client which are not async so I can't make the method async. What do you recommend in this case?

Reply:
You store the client in `hass.data[DOMAIN]`. Can you access name there ?

Reply:
I have update the service parameters to specify entry name instead of host (same as the `transmission` integration). Yes i can get the name from entry.data.

Reply:
Okay. Don't call self.hass.config_entries.async_entries then.

Reply:
We need to find the entry because we are using the same service name for multiple instances. In case the user has configured multiple deluge instances they will need to specify the instance name when calling the service.
This is the main reason for requiring the instance name parameter in the service call.

Reply:
Then store the config entry in `hass.data[DOMAIN]`. You don't even know if the config entries you're iterating over have been set up .

Reply:
The client is stored in hass.data[DOMAIN][entry_id]. I am getting the entry_id from the instance name to get which client to use for the service call.
The same has been implemented in the transmission integration. Please check it to see that its the same.

Reply:
Then transmission is wrong too. 

- You cannot call async methods from sync context
- You cannot look at all avialable config entries because you cannot assume that they are loaded
- You need to store the config entry in `hass.data[DOMAIN]` if that's data that you need later.

Reply:
I would like your help in better understanding this as I assumed that what I did was correct because it was approved in the transmission integration. First let us clarify the following:
- This method is being used by multiple instances of the same integration (`Deluge`) so when calling it we need a way to identify which instance should be used.
- To my understanding `self.hass.config_entries.async_entries(DOMAIN)` method is used to retrieve the config entries loaded for a DOMAIN. `add_torrent` method is related to the service created, it will only be called if there is at least one entry configured and stored so this method won't be called in case no entries are loaded.
- I don't understand why we need to store the config entry in `hass.data[DOMAIN]`. I only need the entry_id to identify which instance I need to use. The instances (or clients) are stored in `hass.data[DOMAIN]` as i explained earlier.
- If calling an async method from sync context is not correct please advise how can we edit this method to make it work with sync and async context (knowing that `Deluge` api is not async)

Appreciate your suggestions to make this better.